File: Spa.API/Spa.API/Spa.API/Entities/LichHen.cs
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace Spa.API.Entities;

[Table("LichHen")]
public class LichHen
{
    [Key]
    public int Id { get; set; }

    [Required, MaxLength(200)]
    public string HoTenKhach { get; set; } = null!;

    [Required, MaxLength(20)]
    public string DienThoaiKhach { get; set; } = null!;

    [MaxLength(100)]
    public string? EmailKhach { get; set; }

    public int? KhachHangId { get; set; }
    public KhachHang? KhachHang { get; set; }

    public DateOnly NgayHen { get; set; }
    public TimeOnly GioHen { get; set; }

    public int ThoiLuongDuKien { get; set; }

    public int? NhanVienId { get; set; }

    public int DichVuId { get; set; }
    public DichVu? DichVu { get; set; }

    [MaxLength(500)]
    public string? GhiChuKhach { get; set; }

    [MaxLength(1000)]
    public string? GhiChuNoiBo { get; set; }

    [Required, MaxLength(50)]
    public string TrangThai { get; set; } = "Chờ xác nhận";

    [MaxLength(500)]
    public string? LyDoHuy { get; set; }

    public DateTime CreatedAt { get; set; }
    public DateTime? UpdatedAt { get; set; }

    [MaxLength(128)]
    public string? CreatedBy { get; set; }

    [MaxLength(128)]
    public string? UpdatedBy { get; set; }

    public bool DaDen { get; set; }
}

File: admin-spa/src/api/lichHen.ts
import { api } from "../lib/api";
import { useCustomerAuthStore } from "../customer/store/auth";

export type LichHenAvailabilityDay = {
  date: string; // DateOnly => "YYYY-MM-DD"
  slots: string[]; // TimeOnly => usually "HH:mm:ss" (or "HH:mm")
};

export type CustomerBookingRow = {
  id: number;
  ngayHen: string;
  gioHen: string;
  thoiLuongDuKien: number;
  trangThai: string;
  dichVuId: number;
  tenDichVu?: string | null;
  hoTenKhach: string;
  dienThoaiKhach: string;
  emailKhach?: string | null;
  ghiChuKhach?: string | null;
  lyDoHuy?: string | null;
  createdAt: string;
  updatedAt?: string | null;
  daDen: boolean;
};

function customerAuthHeaders() {
  const token = useCustomerAuthStore.getState().token;
  return token ? { Authorization: `Bearer ${token}` } : undefined;
}

export const lichHenApi = {
  available: async (params?: { from?: string; days?: number; slotMinutes?: number }) => {
    const res = await api.get<LichHenAvailabilityDay[]>("/api/lich-hen/available", {
      params: {
        from: params?.from,
        days: params?.days ?? 14,
        slotMinutes: params?.slotMinutes ?? 30,
      },
    });
    return res.data;
  },
  create: async (body: {
    hoTenKhach: string;
    dienThoaiKhach: string;
    emailKhach?: string;
    ngayHen: string; // YYYY-MM-DD
    gioHen: string; // HH:mm:ss or HH:mm
    dichVuId: number;
    thoiLuongDuKien: number; // minutes
    ghiChuKhach?: string;
  }) => {
    const res = await api.post<any>("/api/lich-hen", body, {
      headers: customerAuthHeaders(),
    });
    return res.data;
  },
  my: async () => {
    const res = await api.get<CustomerBookingRow[]>("/api/lich-hen/my", {
      headers: customerAuthHeaders(),
    });
    return res.data;
  },
  // staff endpoints (require Bearer token)
  list: async (params?: { from?: string; to?: string; status?: string; nhanVienId?: number; take?: number }) => {
    const res = await api.get<any[]>('/api/lich-hen', {
      params: {
        from: params?.from,
        to: params?.to,
        status: params?.status,
        nhanVienId: params?.nhanVienId,
        take: params?.take ?? 200,
      },
    });
    return res.data;
  },
  confirm: async (id: number, body: { nhanVienId?: number | null; ghiChuNoiBo?: string }) => {
    const payload: { nhanVienId?: number; ghiChuNoiBo?: string } = {};
    if (typeof body.nhanVienId === "number" && body.nhanVienId > 0) payload.nhanVienId = body.nhanVienId;
    if (body.ghiChuNoiBo) payload.ghiChuNoiBo = body.ghiChuNoiBo;
    const res = await api.put<any>(`/api/lich-hen/${id}/confirm`, payload);
    return res.data;
  },
  complete: async (id: number) => {
    const res = await api.put<any>(`/api/lich-hen/${id}/complete`);
    return res.data;
  },
};
